File: src/swarm/core/task.py
"""Task model for swarm orchestration."""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from enum import Enum

from pydantic import BaseModel, Field


class TaskStatus(str, Enum):
    PENDING = "pending"
    ASSIGNED = "assigned"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class Task(BaseModel):
    """A unit of work assigned to an agent."""

    id: str = Field(default_factory=lambda: uuid.uuid4().hex[:12])
    parent_id: str | None = None
    title: str
    description: str = ""
    status: TaskStatus = TaskStatus.PENDING
    assigned_to: str | None = None
    created_by: str = "human"
    result: str | None = None
    error: str | None = None
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    def assign(self, agent_name: str) -> None:
        self.assigned_to = agent_name
        self.status = TaskStatus.ASSIGNED
        self._touch()

    def start(self) -> None:
        self.status = TaskStatus.IN_PROGRESS
        self._touch()

    def complete(self, result: str) -> None:
        self.result = result
        self.status = TaskStatus.COMPLETED
        self._touch()

    def fail(self, error: str) -> None:
        self.error = error
        self.status = TaskStatus.FAILED
        self._touch()

    def cancel(self) -> None:
        self.status = TaskStatus.CANCELLED
        self._touch()

    def _touch(self) -> None:
        self.updated_at = datetime.now(timezone.utc)

File: src/lib/task.ts
import { randomBytes } from 'crypto';

export type TaskStatus = 'pending' | 'assigned' | 'in_progress' | 'completed' | 'failed' | 'cancelled';

export interface TaskData {
  id: string;
  parent_id?: string;
  title: string;
  description: string;
  status: TaskStatus;
  assigned_to?: string;
  created_by: string;
  result?: string;
  error?: string;
  created_at: string;
  updated_at: string;
}

export class Task {
  readonly id: string;
  parent_id?: string;
  title: string;
  description: string;
  status: TaskStatus;
  assigned_to?: string;
  created_by: string;
  result?: string;
  error?: string;
  created_at: string;
  updated_at: string;

  constructor(opts: { title: string; assigned_to?: string; created_by?: string; description?: string }) {
    this.id = randomBytes(6).toString('hex');
    this.title = opts.title;
    this.description = opts.description ?? '';
    this.status = 'pending';
    this.assigned_to = opts.assigned_to;
    this.created_by = opts.created_by ?? 'human';
    this.created_at = new Date().toISOString();
    this.updated_at = new Date().toISOString();
  }

  assign(agentName: string): void { this.assigned_to = agentName; this.status = 'assigned'; this._touch(); }
  start(): void { this.status = 'in_progress'; this._touch(); }
  complete(result: string): void { this.result = result; this.status = 'completed'; this._touch(); }
  fail(error: string): void { this.error = error; this.status = 'failed'; this._touch(); }
  cancel(): void { this.status = 'cancelled'; this._touch(); }

  toJSON(): TaskData {
    return {
      id: this.id, parent_id: this.parent_id, title: this.title,
      description: this.description, status: this.status,
      assigned_to: this.assigned_to, created_by: this.created_by,
      result: this.result, error: this.error,
      created_at: this.created_at, updated_at: this.updated_at,
    };
  }

  static fromJSON(data: TaskData): Task {
    const t = new Task({ title: data.title, assigned_to: data.assigned_to, created_by: data.created_by, description: data.description });
    Object.assign(t, data);
    return t;
  }

  private _touch(): void { this.updated_at = new Date().toISOString(); }
}
